StyleCop.Analyzers/StyleCop.Analyzers.Test/MaintainabilityRules/SA1412UnitTests.cs

```diff
@@ -0,0 +1,139 @@
+﻿namespace StyleCop.Analyzers.Test.MaintainabilityRules
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CodeFixes;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.Text;
+    using StyleCop.Analyzers.MaintainabilityRules;
+    using TestHelper;
+    using Xunit;
+
+    public class SA1412UnitTests : CodeFixVerifier
+    {
+        private Encoding fileEncoding;
+
+        public static IEnumerable<Encoding> NonUtf8Encodings
+        {
+            get
+            {
+                yield return Encoding.ASCII;
+                yield return Encoding.BigEndianUnicode;
+                yield return Encoding.Default;
+                yield return Encoding.Unicode;
+                yield return Encoding.UTF32;
+                yield return Encoding.UTF7;
+            }
+        }
+
+        [Fact]
+        public async Task TestFileWithWrongEncodingAsync()
+        {
+            // Encoding is not IConvertable so xunit does not support it in theories
+            foreach (var encoding in NonUtf8Encodings)
+            {
+                string testCode = "class Foo { }";
+
+                this.fileEncoding = encoding;
+
+                var expected = this.CSharpDiagnostic().WithLocation(1, 1);
+
+                await this.VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None).ConfigureAwait(false);
+            }
+        }
+
+        [Fact]
+        public async Task TestFileWithUtf8EncodingWithoutBOMAsync()
+        {
+            string testCode = "class Foo { }";
+
+            this.fileEncoding = new UTF8Encoding(false);
+
+            var expected = this.CSharpDiagnostic().WithLocation(1, 1);
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None).ConfigureAwait(false);
+        }
+
+        [Fact]
+        public async Task TestCodeFixAsync()
+        {
+            // Encoding is not IConvertable so xunit does not support it in theories
```

Comment:
:bulb: But it would support the code page as an `int`, which you pass to [`Encoding.GetEncoding(int)`](https://msdn.microsoft.com/en-us/library/wzsz3bk3(v=vs.110).aspx) to get the actual encoding to use in the test.